gtsam/linear/SubgraphBuilder.cpp

```diff
@@ -127,7 +127,8 @@ static vector<size_t> UniqueSampler(const vector<double> &weight,
 
     /* sampling and cache results */
     vector<size_t> samples = iidSampler(localWeights, n - count);
-    for (const size_t &index : samples) {
+    const auto samplesSize = samples.size();
+    for (size_t index = 0; index < samplesSize; index++) {
```

Comment:
This is suspicious. To me (a) line 130 should not be an issue for any c++11 compiler, and (b) you are actually changing the behavior, from the *content* of samples to the indices 0...sampleSize-1.

Reply:
The `push_back` inside the for loop sometimes will trigger a reallocation of the vector, and subsequently invalidating all existing iterators.

https://stackoverflow.com/questions/35938329/c-push-back-in-stdvector-while-iterating-it

Reply:
So the `SEGSEGV` will be on line 131 because `index` is invalid.

Reply:
Yikes, the problem is much bigger! This function is completely wrong, as we're shadowing samples on line 110 with new definition on line 129. I will look at PR where this happened to find correct behavior.

Reply:
For reference, PR is on Bitbucket: https://bitbucket.org/gtborg/gtsam/pull-requests/415/split-off-subgraphbuilder/diff

Reply:
Could you restore to `for ( const size_t &index: samples ) {` but change first `samples` to `result` ? push back should happen to `result` and function should return `result`...

Reply:
What is the purpose of `UniqueSampler`? Also the `samples` defined on line 110 is never used.

Reply:
Clearly we need a unit test here to fail, but unit tests were not abundant when this code was created, though I added some in refactor.

Reply:
Note, I have a branch out to make IID sampling better and replace this code, but CI tests failed for some config and I did not get to fixing it yet.

Reply:
> What is the purpose of `UniqueSampler`? Also the `samples` defined on line 110 is never used.

Yeah, that's what I'm saying. In the refactor I renamed result to samples without seeing the shadowing. It should be restored to it's original name.

Reply:
Done :)